Merge ForecastFactory into CercForecastService

jdudley1123 opened the pull request:
## Context
Currently we have two classes, the CercForecastService and the ForecastFactory. CercForecastService is responsible for returning the forecasts from the cache, and refreshing the cache when it is stale. It calls the CERC API and then passes the response to the ForecastFactory.

CercForecastService is the only user of ForecastFactory, and is a very small class, so there is no good reason to have separate class for what is essentially core functionality of the service – building a forecast.

## Changes in this PR
This PR merges ForecastFactory into CercForecastService. It also improves the tests for CercForecastService so that they test behaviour and not implementation.

jdudley1123 committed:
Combine ForecastFactory with CercForecastService

Currently we have two classes, the CercForecastService and the ForecastFactory. CercForecastService is responsible for returning the forecasts from the cache, and refreshing the cache when it is stale. It calls the CERC API and then passes the response to the ForecastFactory.

CercForecastService is the only user of ForecastFactory, and is a very small class, so there is no good reason to have separate class for what is essentially core functionality of the service – building a forecast.

This commit merges ForecastFactory into CercForecastService.

patrickjfl commented on spec/services/cerc_forecast_service_spec.rb:
```diff
@@ -1,6 +1,6 @@
 RSpec.describe CercForecastService do
   describe "::latest_forecasts_for(zone)" do
-    let(:zone) { FactoryBot.build(:zone, name: "Barnet") }
+    let(:zone) { create(:zone, cerc_id: 29) }
```
if we're being consistent here, should we leave it as `FactoryBot.build`?

jdudley1123 replied:
Good spot!

patrickjfl reviewed (approved):
Nice one bringing ForecastFactory into the service 👍 

I can see the benefits of testing behaviour over implementation, but also the drawbacks. Does it not make our tests harder to debug when something changes, causing them to fail?

jdudley1123 committed:
Refactor CercForecastServiceSpec

As a general rule we should test behaviour not implementation. That means we should be testing what the cached forecast is, not which methods were called along the way.

The exception to this is the call to the API because we have stubbed the API.

jdudley1123 commented:
> Nice one bringing ForecastFactory into the service 👍
> 
> I can see the benefits of testing behaviour over implementation, but also the drawbacks. Does it not make our tests harder to debug when something changes, causing them to fail?

If your tests are tightly coupled to your implementation, then you have to change them in lockstep and it becomes difficult to tell whether the change you made has changed the app's behaviour because you've also changed the tests that check that. If the tests are focused on behaviour, then changing the implementation (for example to use a new algorithm) should not mean there are any significant changes to the tests.

Here's a good article about it that actually mentions caches as an exception: https://testing.googleblog.com/2013/08/testing-on-toilet-test-behavior-not.html
